libs/imageio/src/BasisEncoder.cpp

```diff
@@ -0,0 +1,267 @@
+    if (enabled) {
+        mImpl->params.m_tex_type = basist::cBASISTexTypeCubemapArray;
+    } else {
+        const bool multi = mImpl->params.m_source_images.size() > 1;
+        mImpl->params.m_tex_type = multi ? basist::cBASISTexType2DArray : basist::cBASISTexType2D;
+    }
+    return *this;
+}
+
+Builder& Builder::intermediateFormat(BasisEncoder::IntermediateFormat format) noexcept {
+    mImpl->params.m_uastc = format == IntermediateFormat::UASTC;
+    return *this;
+}
+
+Builder& Builder::grayscale(bool enabled) noexcept {
+    mImpl->grayscale = enabled;
+    return *this;
+}
+
+Builder& Builder::normals(bool enabled) noexcept {
+    mImpl->normals = enabled;
+    return *this;
+}
+
+Builder& Builder::jobs(size_t count) noexcept {
+    mImpl->jobs = count;
+    return *this;
+}
+
+Builder& Builder::quiet(bool enabled) noexcept {
+    mImpl->quiet = enabled;
+    return *this;
+}
+
+BasisEncoder* Builder::build() {
+    if (mImpl->error) {
+        return nullptr;
+    }
+
+    basisu::basisu_encoder_init();
+
+    auto& params = mImpl->params;
+
+    params.m_status_output = !mImpl->quiet;
+    params.m_pJob_pool = new basisu::job_pool(mImpl->jobs);
+    params.m_create_ktx2_file = true;
+    params.m_ktx2_uastc_supercompression = basist::KTX2_SS_ZSTANDARD;
+
+    // This sRGB flag doesn't actually affect the compression scheme or the basis format, it's just
+    // an annotation that gets stored in the KTX2 file, which enables the app to choose the right
+    // format when it loads and transcodes the texture. Technically however, the transcoder
+    // SHOULD know about this, since in some scenarios it needs to interpolate between colors.
+    params.m_ktx2_srgb_transfer_func = !mImpl->linear;
+
+    // Select the same quality that the basis tool selects by default (midpoint of range).
+    params.m_quality_level = 128;
+
+    // This is the default zstd compression level used by the basisu cmdline cool.
+    params.m_ktx2_zstd_supercompression_level = 6;
+
+    // We do not want basis to read from files, we want it to read from "m_source_images"
+    params.m_read_source_images = false;
+
+    // We do not want basis to write the file, we want to manually dump "get_output_ktx2_file()"
+    params.m_write_output_basis_files = false;
+
+    basisu::basis_compressor* encoder = new basisu::basis_compressor();
+
+    if (!encoder->init(params)) {
+        assert_invariant(false);
+        delete encoder;
+        return nullptr;
+    }
+
+    return new BasisEncoder(new BasisEncoderImpl {
+        .encoder = encoder,
+        .jobs = params.m_pJob_pool,
+        .quiet = mImpl->quiet,
+    });
+}
+
+BasisEncoder::BasisEncoder(BasisEncoderImpl* impl) noexcept : mImpl(impl) {}
+
+BasisEncoder::~BasisEncoder() noexcept {
+    delete mImpl->encoder;
+    delete mImpl->jobs;
+    delete mImpl;
+    basisu::basisu_encoder_deinit();
+}
+
+BasisEncoder::BasisEncoder(BasisEncoder&& that) noexcept {
```

Comment:
I think you need to initialize `mImpl` here, otherwise it doesn't get initialized.

Reply:
Addressed this and the other items in the latest commit.